pinterest/gestalt
# ComboBox: new component

The current Gestalt  Typeahead component is buggy and lacking several needed features. Rather than add to our tech debt, we rewrite ComboBox from scratch and implement a subset of the feature requests along the way.

ComboBox is the combination of a Textfield and an associated Dropdown to let the user filter a list when selecting. ComboBox allows to type the selection, type part of the selection and narrow the results, and select an option from the list.

![image](https://user-images.githubusercontent.com/10593890/123166450-6bb73c00-d443-11eb-8880-0cf57b7b17f4.png)
![image](https://user-images.githubusercontent.com/10593890/123166470-7376e080-d443-11eb-9c0f-9dbbef3d8e32.png)


https://www.figma.com/file/vjhfBsOtHw0wVg67vqwz1v/01.-Web-Sticker-Sheet?node-id=2508%3A27703

## Description

How to manually test this component

- Go through all examples particularly in controlled / uncontrolled / tags
- Select an option with mouse
- Reselect an option with mouse
- Type new value to narrow down options
- Clean typed values using `delete`
- Clean typed values using `clear` button
- Use keyboard navigation to get to combobox, open options list, navigate through options, select an option pressing ENTER, reselect another option using keyboard navigation, move to next interactive element
- Try as many iterations of the above steps to make sure all possible behavior work correctly.
- Try darkmode, RTL
- Set disable state in regular ComboBox and with tags
- Set a helper text and error message
- Set screenreader and navigate through the ComboBox

### Links
- [TDD](https://paper.dropbox.com/doc/Gestalt-TDD-ComboBox--BNOsc9byxy8GmClOKiIKd~B6Ag-nMa6qqKQPEvjmVkvPzfgJ)

### Checklist

- [x] Added Unit and Flow Tests
- [x] Added documentation + accessibility tests
- [x] Verified Accessibility: keyboard & screen reader interaction
- [x] Checked Dark Mode, Responsiveness, and Right-to-Left support
- [x] Checked Stakeholder feedback (ex. Gestalt designers)
